File: EPI/13.4/swift/13.4.swift
private struct CachedValue {
    let price: Int
    var olderISBN: String?
    var newerISBN: String?
}

public struct ISBNCache {
    public let capacity: Int
    private var cache: [String: CachedValue]
    private var oldestISBN: String?
    private var newestISBN: String?

    public init(capacity: Int) {
        self.capacity = capacity
        self.cache = [String: CachedValue](minimumCapacity: capacity)
    }

    public mutating func lookup(_ ISBN: String) -> Int? {
        if let price = cache[ISBN]?.price {
            // Removing then inserting is the easiest way to bump the ISBN to newest, but it also
            // performs unnecessary work and could be optimized.
            self.remove(ISBN)
            self.insert(ISBN, price: price)
            return price
        }

        return nil
    }

    public mutating func insert(_ ISBN: String, price: Int) {
        if cache[ISBN] != nil {
            return
        }

        // Remove oldest ISBN if necessary.
        if cache.count >= capacity {
            self.remove(self.oldestISBN!)
        }

        // Insert as newest ISBN.
        cache[ISBN] = CachedValue(price: price, olderISBN: self.newestISBN, newerISBN: nil)
        if let newestISBN = self.newestISBN {
            cache[newestISBN]!.newerISBN = ISBN
        }
        self.newestISBN = ISBN
        self.oldestISBN = self.oldestISBN ?? ISBN
    }

    public mutating func remove(_ ISBN: String) {
        if let value = cache[ISBN] {
            cache.removeValue(forKey: ISBN)

            if let olderISBN = value.olderISBN {
                cache[olderISBN]!.newerISBN = value.newerISBN
            }

            if let newerISBN = value.newerISBN {
                cache[newerISBN]!.olderISBN = value.olderISBN
            }

            if ISBN == self.oldestISBN {
                self.oldestISBN = value.newerISBN
            }

            if ISBN == self.newestISBN {
                self.newestISBN = value.olderISBN
            }
        }
    }
}

File: EPI/Hash Tables/Implement an ISBN cache/typescript/13.4.ts
interface Node {
    ISBN: string;
    prev?: Node;
    next?: Node;
}

export default class ISBNCache {
    private first: Node | undefined;
    private last: Node | undefined;
    private map = new Map<string, {node: Node, price: number}>();

    constructor(public capacity: number) {}

    insert(ISBN: string, price: number): void {
        this.remove(ISBN);

        if (this.map.size >= this.capacity) {
            this.remove(this.last!.ISBN);
        }

        const node: Node = {ISBN, next: this.first};
        this.map.set(ISBN, {node, price});

        if (this.first) {
            this.first.prev = node;
        }

        this.first = node;

        if (!this.last) {
            this.last = node;
        }
    }

    lookup(ISBN: string): number | undefined {
        if (!this.map.has(ISBN)) {
            return undefined;
        }

        const {price} = this.map.get(ISBN)!;
        this.remove(ISBN);
        this.insert(ISBN, price);
        return price;
    }

    remove(ISBN: string): void {
        if (!this.map.has(ISBN)) {
            return;
        }

        const {node} = this.map.get(ISBN)!;
        this.map.delete(ISBN);

        if (node === this.first) {
            this.first = node.next;
        }

        if (node === this.last) {
            this.last = node.prev;
        }

        if (node.prev) {
            node.prev.next = node.next;
        }

        if (node.next) {
            node.next.prev = node.prev;
        }
    }
}
